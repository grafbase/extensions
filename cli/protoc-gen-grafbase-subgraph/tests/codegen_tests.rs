#![allow(clippy::panic)]

use std::{collections::BTreeMap, fs, path::Path, process};

/// Parse a test file containing multiple proto files marked with //!file: comments
fn parse_test_file(content: &str) -> BTreeMap<String, String> {
    let mut files = BTreeMap::new();
    let mut current_file = None;
    let mut current_content = String::new();

    for line in content.lines() {
        if let Some(filename) = line.strip_prefix("//!file:") {
            // Save previous file if any
            if let Some(file) = current_file.take() {
                files.insert(file, current_content.trim().to_string());
                current_content.clear();
            }
            current_file = Some(filename.trim().to_string());
        } else if current_file.is_some() {
            current_content.push_str(line);
            current_content.push('\n');
        }
    }

    // Save last file
    if let Some(file) = current_file {
        files.insert(file, current_content.trim().to_string());
    }

    files
}

/// Write parsed proto files to a temporary directory
fn setup_proto_files(tmp_dir: &Path, files: &BTreeMap<String, String>) -> Vec<String> {
    let mut main_protos = Vec::new();

    for (filename, content) in files {
        let file_path = tmp_dir.join(filename);

        // Create parent directories if needed
        if let Some(parent) = file_path.parent() {
            fs::create_dir_all(parent).unwrap();
        }

        fs::write(&file_path, content).unwrap();

        // Track main proto files (not in subdirectories like grafbase/options.proto)
        if !filename.contains('/') || !filename.starts_with("grafbase/") {
            main_protos.push(filename.clone());
        }
    }

    main_protos
}

#[test]
fn codegen_tests() {
    let bin_path = env!("CARGO_BIN_EXE_protoc-gen-grafbase-subgraph");

    if cfg!(windows) {
        eprintln!("Skipping these tests on Windows, since there is no official protoc binary release.");
        return;
    }

    insta::glob!("testcases/*.proto", |test_path| {
        let snapshot_path = test_path.parent().unwrap();
        let test_name = test_path.file_stem().unwrap().to_str().unwrap();

        insta::with_settings!({
            snapshot_path => snapshot_path.to_str().unwrap(),
            prepend_module_to_snapshot => false,
            snapshot_suffix => "",
        }, {
            let test_content = fs::read_to_string(test_path).unwrap();
            let mut files = parse_test_file(&test_content);

            // If no //!file: markers, treat as single file for backwards compatibility
            if files.is_empty() {
                files.insert(
                    test_path.file_name().unwrap().to_str().unwrap().to_string(),
                    test_content,
                );
            }

            // Always insert the options.proto from the source of truth
            // This will override any inline options.proto in test files
            let options_proto_path = Path::new(env!("CARGO_MANIFEST_DIR"))
                .join("proto")
                .join("grafbase")
                .join("options.proto");
            let options_content = fs::read_to_string(options_proto_path).unwrap();
            files.insert("grafbase/options.proto".to_string(), options_content);

            let proto_tmp = tempfile::tempdir().unwrap();
            let main_protos = setup_proto_files(proto_tmp.path(), &files);

            let output_tmp = tempfile::tempdir().unwrap();

            for main_proto in main_protos {
                let mut cmd = process::Command::new("protoc");

                cmd.arg("--plugin")
                    .arg(bin_path)
                    .arg("--grafbase-subgraph_out")
                    .arg(output_tmp.path())
                    .arg("-I")
                    .arg(proto_tmp.path())
                    .arg(&main_proto)
                    .stderr(process::Stdio::inherit());

                let output = cmd.output().unwrap();

                assert!(
                    output.status.success(),
                    "Expected success for {}, got {}\n{}",
                    main_proto,
                    output.status,
                    std::str::from_utf8(&output.stderr).unwrap(),
                );
            }

            // Collect all generated .graphql files
            let mut graphql_files = BTreeMap::new();
            for entry in walkdir::WalkDir::new(output_tmp.path()) {
                let entry = entry.unwrap();

                if !entry.file_type().is_file() {
                    continue;
                }

                if let Some("graphql") = entry.path().extension().and_then(|s| s.to_str()) {
                    let filename = entry.file_name().to_str().unwrap().to_string();
                    let content = fs::read_to_string(entry.path()).unwrap();
                    graphql_files.insert(filename, content);
                }
            }

            // Debug: print what files were generated
            if test_name.contains("multi") {
                eprintln!("Generated files for {}: {:?}", test_name, graphql_files.keys().collect::<Vec<_>>());
            }

            // Concatenate all files in deterministic order
            let mut combined_output = String::new();
            for (filename, content) in graphql_files {
                if !combined_output.is_empty() {
                    combined_output.push_str("\n#--- ");
                    combined_output.push_str(&filename);
                    combined_output.push_str(" ---#\n\n");
                } else {
                    combined_output.push_str("#--- ");
                    combined_output.push_str(&filename);
                    combined_output.push_str(" ---#\n\n");
                }
                combined_output.push_str(&content);
            }

            insta::assert_snapshot!(test_name, combined_output);
        });
    });
}
