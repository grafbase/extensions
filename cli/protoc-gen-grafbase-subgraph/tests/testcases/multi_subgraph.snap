---
source: cli/protoc-gen-grafbase-subgraph/tests/codegen_tests.rs
expression: combined_output
input_file: cli/protoc-gen-grafbase-subgraph/tests/testcases/multi_subgraph.proto
---
#--- orders.graphql ---#

extend schema
  @link(url: "https://grafbase.com/extensions/grpc/0.2.0", import: ["@protoServices", "@protoEnums", "@protoMessages", "@grpcMethod"])
  @protoServices(
    definitions: [
      {
        name: "example.OrderService"
        methods: [
          {
            name: "CreateOrder"
            inputType: ".example.CreateOrderRequest"
            outputType: ".example.Order"
          }
          {
            name: "GetOrder"
            inputType: ".example.GetOrderRequest"
            outputType: ".example.Order"
          }
        ]
      }
    ]
  )
  @protoMessages(
    definitions: [
      {
        name: ".example.Order"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "user_id"
            number: 2
            repeated: false
            type: "string"
          }
          {
            name: "product_ids"
            number: 3
            repeated: true
            type: "string"
          }
          {
            name: "total"
            number: 4
            repeated: false
            type: "double"
          }
        ]
      }
      {
        name: ".example.GetOrderRequest"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".example.CreateOrderRequest"
        fields: [
          {
            name: "user_id"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "product_ids"
            number: 2
            repeated: true
            type: "string"
          }
        ]
      }
    ]
  )

type Mutation {
  example_OrderService_CreateOrder(input: example_CreateOrderRequestInput): example_Order @grpcMethod(service: "example.OrderService", method: "CreateOrder")
  example_OrderService_GetOrder(input: example_GetOrderRequestInput): example_Order @grpcMethod(service: "example.OrderService", method: "GetOrder")
}

"64 bit signed integer" scalar I64
"64 bit unsigned integer" scalar U64

input example_GetOrderRequestInput {
  id: String
}

input example_CreateOrderRequestInput {
  user_id: String
  product_ids: [String!]
}

type example_Order {
  id: String!
  user_id: String!
  product_ids: [String!]
  total: Float!
}

#--- products.graphql ---#

extend schema
  @link(url: "https://grafbase.com/extensions/grpc/0.2.0", import: ["@protoServices", "@protoEnums", "@protoMessages", "@grpcMethod"])
  @protoServices(
    definitions: [
      {
        name: "example.ProductService"
        methods: [
          {
            name: "GetProduct"
            inputType: ".example.GetProductRequest"
            outputType: ".example.Product"
          }
          {
            name: "CreateProduct"
            inputType: ".example.CreateProductRequest"
            outputType: ".example.Product"
          }
          {
            name: "ListProducts"
            inputType: ".example.ListProductsRequest"
            outputType: ".example.ListProductsResponse"
          }
        ]
      }
    ]
  )
  @protoMessages(
    definitions: [
      {
        name: ".example.Product"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "name"
            number: 2
            repeated: false
            type: "string"
          }
          {
            name: "price"
            number: 3
            repeated: false
            type: "double"
          }
          {
            name: "owner_id"
            number: 4
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".example.GetProductRequest"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".example.CreateProductRequest"
        fields: [
          {
            name: "name"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "price"
            number: 2
            repeated: false
            type: "double"
          }
          {
            name: "owner_id"
            number: 3
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".example.ListProductsRequest"
        fields: [
          {
            name: "limit"
            number: 1
            repeated: false
            type: "int32"
          }
          {
            name: "cursor"
            number: 2
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".example.ListProductsResponse"
        fields: [
          {
            name: "products"
            number: 1
            repeated: true
            type: ".example.Product"
          }
          {
            name: "next_cursor"
            number: 2
            repeated: false
            type: "string"
          }
        ]
      }
    ]
  )

type Mutation {
  example_ProductService_GetProduct(input: example_GetProductRequestInput): example_Product @grpcMethod(service: "example.ProductService", method: "GetProduct")
  example_ProductService_CreateProduct(input: example_CreateProductRequestInput): example_Product @grpcMethod(service: "example.ProductService", method: "CreateProduct")
  example_ProductService_ListProducts(input: example_ListProductsRequestInput): example_ListProductsResponse @grpcMethod(service: "example.ProductService", method: "ListProducts")
}

"64 bit signed integer" scalar I64
"64 bit unsigned integer" scalar U64

input example_GetProductRequestInput {
  id: String
}

input example_CreateProductRequestInput {
  name: String
  price: Float
  owner_id: String
}

input example_ListProductsRequestInput {
  limit: Int
  cursor: String
}

type example_Product {
  id: String!
  name: String!
  price: Float!
  owner_id: String!
}

type example_ListProductsResponse {
  products: [example_Product!]
  next_cursor: String!
}

#--- users.graphql ---#

extend schema
  @link(url: "https://grafbase.com/extensions/grpc/0.2.0", import: ["@protoServices", "@protoEnums", "@protoMessages", "@grpcMethod"])
  @protoServices(
    definitions: [
      {
        name: "example.UserService"
        methods: [
          {
            name: "GetUser"
            inputType: ".example.GetUserRequest"
            outputType: ".example.User"
          }
          {
            name: "CreateUser"
            inputType: ".example.CreateUserRequest"
            outputType: ".example.User"
          }
          {
            name: "ListUsers"
            inputType: ".example.ListUsersRequest"
            outputType: ".example.ListUsersResponse"
          }
        ]
      }
    ]
  )
  @protoMessages(
    definitions: [
      {
        name: ".example.User"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "name"
            number: 2
            repeated: false
            type: "string"
          }
          {
            name: "email"
            number: 3
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".example.GetUserRequest"
        fields: [
          {
            name: "id"
            number: 1
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".example.CreateUserRequest"
        fields: [
          {
            name: "name"
            number: 1
            repeated: false
            type: "string"
          }
          {
            name: "email"
            number: 2
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".example.ListUsersRequest"
        fields: [
          {
            name: "limit"
            number: 1
            repeated: false
            type: "int32"
          }
          {
            name: "cursor"
            number: 2
            repeated: false
            type: "string"
          }
        ]
      }
      {
        name: ".example.ListUsersResponse"
        fields: [
          {
            name: "users"
            number: 1
            repeated: true
            type: ".example.User"
          }
          {
            name: "next_cursor"
            number: 2
            repeated: false
            type: "string"
          }
        ]
      }
    ]
  )

type Mutation {
  example_UserService_GetUser(input: example_GetUserRequestInput): example_User @grpcMethod(service: "example.UserService", method: "GetUser")
  example_UserService_CreateUser(input: example_CreateUserRequestInput): example_User @grpcMethod(service: "example.UserService", method: "CreateUser")
  example_UserService_ListUsers(input: example_ListUsersRequestInput): example_ListUsersResponse @grpcMethod(service: "example.UserService", method: "ListUsers")
}

"64 bit signed integer" scalar I64
"64 bit unsigned integer" scalar U64

input example_GetUserRequestInput {
  id: String
}

input example_CreateUserRequestInput {
  name: String
  email: String
}

input example_ListUsersRequestInput {
  limit: Int
  cursor: String
}

"""
Shared types
"""
type example_User {
  id: String!
  name: String!
  email: String!
}

type example_ListUsersResponse {
  users: [example_User!]
  next_cursor: String!
}
